Домашнее задание 1 - 10 баллов
1) Загрузите набор данных lenta-ru-news с помощью библиотеки Corus для задачи классификации текстов по топикам (пригодятся атрибуты title, text, topic)- 1 балл
2) Подготовьте данные к обучению: - 3 балла
3) Предобработайте данные: реализуйте оптимальную, на ваш взгляд, предобработку текстов (нормализация, очистка, стемминг/лемматизация и т.п.) и таргета.
hint: для ускорения обработки и обучения можно ограничиться не всем датасетом, а его репрезентативной частью, например, размера 100_000.
4) Кратко опишите пайплайн, на котором остановились, и почему.
5) Разделите датасет на обучающую, валидационную и тестовую выборки со стратификацией в пропорции 60/20/20. В качестве целевой переменной используйте атрибут topic
6) Замерьте базовое качество с любым dummy-бейзлайном - 0.5 балла
7) Обучите модель sklearn.linear_model.LogisticRegression с двумя вариантами векторизации: 2 балла
sklearn.feature_extraction.text.CountVectorizer
sklearn.feature_extraction.text.TfidfVectorizer
8) Попробуйте улучшить качество, подобрав оптимальные гиперпараметры трансформаций и модели на кросс-валидации 1 балл
9) Оцените качество лучшего пайплайна на отложенной выборке - 0.5 балла

Общее

Принимаемые решения обоснованы (почему выбрана определенная архитектура/гиперпараметр/оптимизатор/преобразование и т.п.) - 1 балл
Обеспечена воспроизводимость решения: зафиксированы random_state, ноутбук воспроизводится от начала до конца без ошибок - 1 балл
Формат сдачи ДЗ

Создать отдельный репозиторий (если еще не создавали)
Выдать доступ в репозиторий pacifikus
Каждая домашняя работа – PR в отдельную ветку hw_n, где n - номер домашней работы
Добавить pacifikus в reviewers
Дождаться ревью, если все ок – мержим в main
Если не ок – вносим исправления и снова отправляем на ревью

# 1. Загрузка данных через Corus, установка и импорт либ

In [1]:
!pip install pandas corus natasha pandarallel requests beautifulsoup4 razdel numpy scikit-learn pymorphy3

  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 83.7/83.7 kB 1.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 34.4/34.4 MB 18.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.8/53.8 kB 3.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.4/119.4 kB 5.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.7/46.7 kB 2.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.5/55.5 kB 3.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.4/8.4 MB 31.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.2/8.2 MB 29.3 MB/s eta 0:00:00
  Created wheel for pandarallel: filename=pandarallel-1.6.5-py3-none-any.whl size=16674 sha256=7307b9d6d16a3668b177b07b296771ea04cd565cbaf2c1ee8f8f3bc363d28867
  Stored in directory: /root/.cache/pip/wheels/b9/c6/5a/829298789e94348b81af52ab42c19d49da00

In [2]:
!wget https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz

--2025-03-11 19:26:45--  https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz
Resolving github.com (github.com)... 140.82.114.3
Connecting to github.com (github.com)|140.82.114.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250311%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250311T192645Z&X-Amz-Expires=300&X-Amz-Signature=11256a38d69198b367392d7ca5e94f0e66a777a59e48f481a4d784610a984d64&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dlenta-ru-news.csv.gz&response-content-type=application%2Foctet-stream [following]
--2025-03-11 19:26:45--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-HMAC-

In [3]:
import pandas as pd
import corus
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.dummy import DummyClassifier
from sklearn.metrics import classification_report, accuracy_score
from sklearn.compose import ColumnTransformer
import pandas as pd
from pymorphy3 import MorphAnalyzer
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from pandarallel import pandarallel
import re
import nltk
import numpy as np
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline

In [4]:
path = 'lenta-ru-news.csv.gz'
records = list(corus.load_lenta(path))
records[0]

LentaRecord(
    url='https://lenta.ru/news/2018/12/14/cancer/',
    title='Названы регионы России с\xa0самой высокой смертностью от\xa0рака',
    text='Вице-премьер по социальным вопросам Татьяна Голикова рассказала, в каких регионах России зафиксирована наиболее высокая смертность от рака, сообщает РИА Новости. По словам Голиковой, чаще всего онкологические заболевания становились причиной смерти в Псковской, Тверской, Тульской и Орловской областях, а также в Севастополе. Вице-премьер напомнила, что главные факторы смертности в России — рак и болезни системы кровообращения. В начале года стало известно, что смертность от онкологических заболеваний среди россиян снизилась впервые за три года. По данным Росстата, в 2017 году от рака умерли 289 тысяч человек. Это на 3,5 процента меньше, чем годом ранее.',
    topic='Россия',
    tags='Общество',
    date=None
)

In [5]:
df = pd.DataFrame(records)
df = df[[1, 2, 3]]
df = df.sample(n=100_000, random_state=1)
print(len(df))
df.head()

100000


,1,2,3
544337,В Южной Осетии задержали российских журналистов,"Съемочная группа программы ""Вести"" телеканала ...",Бывший СССР
683644,Израиль защитил свой ядерный центр от иракских...,Израиль в ночь со среды на четверг разместил п...,Мир
149458,Меховые помпоны стали новым трендом у звезд Го...,Новым популярным трендом в этом сезоне стали м...,Ценности
600335,Буш публично признал наличие секретной програм...,"В субботу президент США Джордж Буш заявил, что...",Мир
311396,Сербский наркобарон заказал убийство бывшего п...,Сербский преступный авторитет Дарко Шарич объя...,Мир


Здесь ничего сложного, выгрузил данные, взял нужные атрибуты: 1, 2, 3. Для усокрения ограничился репрезентативной частью в 100к

# 2-3. Подготовка и предобработка данных

In [6]:
nltk.download('punkt_tab')

nltk.download('stopwords')
nltk.download('punkt')
pandarallel.initialize(progress_bar=True)

morph = MorphAnalyzer()
stop_words_ru = set(stopwords.words('russian'))

def preprocess_text(text):

    clean_text = re.sub(r'[^а-яёА-ЯЁ\s-]', '', str(text), flags=re.IGNORECASE)

    tokens = word_tokenize(clean_text, language='russian')

    processed = [
        morph.parse(token.lower())[0].normal_form
        for token in tokens
        if token.lower() not in stop_words_ru and len(token) > 1
    ]

    return ' '.join(processed)

df_orig = df.copy()
df[1] = df[1].parallel_apply(preprocess_text)
df[2] = df[2].parallel_apply(preprocess_text)

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


INFO: Pandarallel will run on 1 workers.
INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.


Применям условно стандартные преобразования для русского языка:
1. Удаляем цифры и другие спецмимволы
2. Удаляем стоп слова
3. Приводим в норм. форму

### Выгрузка и загрузка df


In [7]:
df.to_csv(
    'lenta-ru-new-ready-pipeline.csv.gz',
    index=False,
    encoding='utf-8',
    compression='gzip'
)

In [8]:
df = pd.read_csv(
    'lenta-ru-new-ready-pipeline.csv.gz',
    compression='gzip',
    encoding='utf-8'
)

In [9]:
df.rename(columns={df.columns[2]: 'topic'}, inplace=True)
df.rename(columns={df.columns[1]: 'text'}, inplace=True)
df.rename(columns={df.columns[0]: 'title'}, inplace=True)

In [10]:
df

,title,text,topic
0,южный осетия задержать российский журналист,съёмочный группа программа вести телеканал рос...,Бывший СССР
1,израиль защитить свой ядерный центр иракский б...,израиль ночь среда четверг разместить противор...,Мир
2,меховой помпон стать новый тренд звезда голливуд,новый популярный тренд сезон стать меховой пом...,Ценности
3,буш публично признать наличие секретный програ...,суббота президент сша джордж буш заявить намер...,Мир
4,сербский наркобарон заказать убийство бывший п...,сербский преступный авторитет дарко шарич объя...,Мир
...,...,...,...
99995,российский милиция планировать сделать федерал...,год российский милиция планироваться финансиро...,Россия
99996,бывший противник буш советовать керри рассчиты...,гарри маурый политик участвовать дебаты джордж...,Мир
99997,явка выборы парламент северный осетия превысит...,явка выборы законодательный собрание северный ...,Россия
99998,остров хоккайдо произойти землетрясение балл,землетрясение сила балл шкала рихтер произойти...,Мир


Чтоб избежать повторной обработки данных, заюзал сохранение df в файл с последующей возможностью юзать его. Да и в целом полезная практика сохранять промежуточные данные от возможных потерь

# 5-9. Разделение данных и бейзлайн, обучение, подбор гиперпараметров, оценка

```



In [11]:
topic_counts = df['topic'].value_counts()
valid_topics = topic_counts[topic_counts >= 500].index
df = df[df['topic'].isin(valid_topics)]

df_full = df.copy()

print("Датасет:")
print(df_full['topic'].value_counts())


Датасет:
topic
Россия               21742
Мир                  18425
Экономика            10827
Спорт                 8767
Культура              7387
Наука и техника       7187
Бывший СССР           7114
Интернет и СМИ        6088
Из жизни              3735
Дом                   2897
Силовые структуры     2667
Ценности              1021
Бизнес                 977
Путешествия            835
Name: count, dtype: int64


Попробовал убрать классы, где значений было <500

In [12]:
df_full['topic'], unique_topics = pd.factorize(df_full['topic'])

In [13]:
df_full

,title,text,topic
0,южный осетия задержать российский журналист,съёмочный группа программа вести телеканал рос...,0
1,израиль защитить свой ядерный центр иракский б...,израиль ночь среда четверг разместить противор...,1
2,меховой помпон стать новый тренд звезда голливуд,новый популярный тренд сезон стать меховой пом...,2
3,буш публично признать наличие секретный програ...,суббота президент сша джордж буш заявить намер...,1
4,сербский наркобарон заказать убийство бывший п...,сербский преступный авторитет дарко шарич объя...,1
...,...,...,...
99995,российский милиция планировать сделать федерал...,год российский милиция планироваться финансиро...,4
99996,бывший противник буш советовать керри рассчиты...,гарри маурый политик участвовать дебаты джордж...,1
99997,явка выборы парламент северный осетия превысит...,явка выборы законодательный собрание северный ...,4
99998,остров хоккайдо произойти землетрясение балл,землетрясение сила балл шкала рихтер произойти...,1


In [14]:
RANDOM_STATE=5235

print("Пропуски в df_full до обработки:")
print(df_full[['title', 'text']].isna().sum())

df_full['title'] = df_full['title'].fillna('')
df_full['text'] = df_full['text'].fillna('')

X_full = df_full['title'] + ' ' + df_full['text']
y_full = df_full['topic']

# Разделяем данных с учетом стратификации, где 60/20/20
X_train_full, X_temp_full, y_train_full, y_temp_full = train_test_split(
    X_full,
    y_full,
    test_size=0.4,
    stratify=y_full,
    random_state=RANDOM_STATE
)

X_val_full, X_test_full, y_val_full, y_test_full = train_test_split(
    X_temp_full,
    y_temp_full,
    test_size=0.5,
    stratify=y_temp_full,
    random_state=RANDOM_STATE
)

print(f"\nРазмеры для df_full:")
print(f"Train: {len(X_train_full)}")
print(f"Val: {len(X_val_full)}")
print(f"Test: {len(X_test_full)}")


# Бейзлайн модель
dummy_full = DummyClassifier(strategy="stratified", random_state=RANDOM_STATE)
dummy_full.fit(X_train_full, y_train_full)

print("\nDummy (full dataset) validation score:", dummy_full.score(X_val_full, y_val_full))
print("Dummy (full dataset) test score:", dummy_full.score(X_test_full, y_test_full))

# Ну и само обучение моделей
def train_evaluate_pipeline(vectorizer, X_tr, y_tr, X_v, y_v):
    pipeline = Pipeline([
        ('vectorizer', vectorizer),
        ('model', LogisticRegression(random_state=RANDOM_STATE, max_iter=500))
    ])

    pipeline.fit(X_tr, y_tr)
    print(f"\n{vectorizer.__class__.__name__} validation accuracy:", pipeline.score(X_v, y_v))
    return pipeline

print("\nОбучаем модели на датасете:")
count_pipe_full = train_evaluate_pipeline(CountVectorizer(), X_train_full, y_train_full, X_val_full, y_val_full)
tfidf_pipe_full = train_evaluate_pipeline(TfidfVectorizer(), X_train_full, y_train_full, X_val_full, y_val_full)

Пропуски в df_full до обработки:
title     0
text     63
dtype: int64

Размеры для df_full:
Train: 59801
Val: 19934
Test: 19934

Dummy (full dataset) validation score: 0.12451088592354771
Dummy (full dataset) test score: 0.12019664894150697

Обучаем модели на датасете:

CountVectorizer validation accuracy: 0.8012441055483094

TfidfVectorizer validation accuracy: 0.8121300290960168


In [16]:
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline

# Пытаемся улучшить
pipeline = Pipeline([
    ('vectorizer', TfidfVectorizer()),
    ('model', LogisticRegression(
        random_state=RANDOM_STATE,
        max_iter=1000,
        class_weight='balanced'
    ))
])

params = {
    'vectorizer__max_features': [1000, 5000],
    'vectorizer__ngram_range': [(1, 1), (1, 2)],
}

grid_search = GridSearchCV(
    pipeline,
    param_grid=params,
    cv=3,
    scoring='accuracy'
)

grid_search.fit(X_train_full, y_train_full)

print("Лучшие параметры:")
print(grid_search.best_params_)

print("\nЛучшая точность на кросс-валидации:", grid_search.best_score_)
print("Точность на валидационной выборке:", grid_search.score(X_val_full, y_val_full))

best_model = grid_search.best_estimator_
y_test_pred = best_model.predict(X_test_full)

print("\nФинальный отчет на тестовых данных:")
print(classification_report(y_test_full, y_test_pred))

Лучшие параметры:
{'vectorizer__max_features': 5000, 'vectorizer__ngram_range': (1, 2)}

Лучшая точность на кросс-валидации: 0.7830303244238488
Точность на валидационной выборке: 0.7895555332597572

Финальный отчет на тестовых данных:
              precision    recall  f1-score   support

           0       0.78      0.87      0.82      1423
           1       0.83      0.76      0.79      3685
           2       0.74      0.82      0.78       204
           3       0.86      0.80      0.82      2165
           4       0.86      0.69      0.77      4349
           5       0.45      0.74      0.56       534
           6       0.81      0.80      0.81      1437
           7       0.96      0.96      0.96      1753
           8       0.49      0.68      0.57       747
           9       0.88      0.89      0.88      1477
          10       0.72      0.76      0.74      1218
          11       0.74      0.89      0.81       580
          12       0.30      0.59      0.39       195
        

# Мощные выводы, мнения, пояснения:


## **bert**:
Изначально не стал брать BERT или RUBERT по причине долгого обучения, на 100к понадобилось бы примерно 5 часов обучения на gpu

## **по предобработке**:
фор рашн лангуяге лемматизация работает точнее чем стемминг

стоп-слова не несут логической нагрузки, поэтому от них можно избавиться, то же самое с неалфавитными символами

title + text даёт контекст для устранения возможных неодназначностей

## **сравнение CountVectorizer vs TF-IDF**
CountVectorizer учитывет частоту слов
TF-IDF cнижает вес частых слов, усиливая значимость редких но возможно информативных терминов

на валидации:
CountVectorizer: accuracy 0.80

TF-IDF: accuracy 0.81

## **гиперпараметры**

учёт униграмм и биграмм увеличвает accuracy

++ балансровка классов предотвращает игнорирование редких классов (вроде как?)

max_features пробовал разные, но из-за того что выставить большие значения просто не получился из-за увеличения времени, выставил в 5к как компромисс


## **метрики и т.д**

Нуууууу в целом норм, соответствует логистической регрессии, видим что accuracy 0.79, есть хорошие результаты. Но также есть и проблемные классы с низким precision. Скорее всего из-за дисбаланса классов либо перекрытия терминов между собой~


## мысли
надо побороть дисбаланс, мб применить синтетическую генерацию примеров для редких классов



